### ISSUE TITLE
Unable to upgrade from Zammad 3.3 to 3.6

### ISSUE THREAD

**Issue by martini** [zammad/zammad#3282]
### Infos:

* Used Zammad version: 3.6.0
* Installation method (source, package, ..): any
* Operating system: any
* Database + version: any
* Elasticsearch version: any
* Browser + version: any

### Expected behavior:

* execute `rake db:migrate` as expected

### Actual behavior:

* `rake db:migrate`

```
rake db:migrate
== 20200527000000 AgentCustomer: migrating ====================================
rake aborted!
StandardError: An error has occurred, this and all later migrations canceled:

undefined method `allow_signup' for #<Permission:0x0000000009c679e8>
/usr/local/rvm/gems/ruby-2.6.6/gems/activemodel-5.2.4.4/lib/active_model/attribute_methods.rb:430:in `method_missing'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/relation/delegation.rb:71:in `each'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/relation/delegation.rb:71:in `each'
/home/xxx/zammad/app/models/role.rb:227:in `reject'
/home/xxx/zammad/app/models/role.rb:227:in `check_default_at_signup_permissions'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:426:in `block in make_lambda'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:198:in `block (2 levels) in halting'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:606:in `block (2 levels) in default_terminator'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:605:in `catch'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:605:in `block in default_terminator'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:199:in `block in halting'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:513:in `block in invoke_before'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:513:in `each'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:513:in `invoke_before'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:131:in `run_callbacks'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:816:in `_run_update_callbacks'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/callbacks.rb:350:in `_update_record'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/timestamp.rb:114:in `_update_record'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/persistence.rb:705:in `create_or_update'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/callbacks.rb:342:in `block in create_or_update'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:132:in `run_callbacks'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:816:in `_run_save_callbacks'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/callbacks.rb:342:in `create_or_update'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/persistence.rb:308:in `save!'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/validations.rb:52:in `save!'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/transactions.rb:315:in `block in save!'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/transactions.rb:387:in `block in with_transaction_returning_status'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/connection_adapters/abstract/database_statements.rb:265:in `transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/transactions.rb:212:in `transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/transactions.rb:385:in `with_transaction_returning_status'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/transactions.rb:315:in `save!'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/suppressor.rb:48:in `save!'
/home/xxx/zammad/db/migrate/20200527000000_agent_customer.rb:8:in `block in change'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/relation/delegation.rb:71:in `each'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/relation/delegation.rb:71:in `each'
/home/xxx/zammad/db/migrate/20200527000000_agent_customer.rb:6:in `change'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:814:in `exec_migration'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:798:in `block (2 levels) in migrate'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:797:in `block in migrate'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:416:in `with_connection'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:796:in `migrate'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:977:in `migrate'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1292:in `block in execute_migration_in_transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1343:in `block in ddl_transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/connection_adapters/abstract/database_statements.rb:267:in `block in transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/connection_adapters/abstract/transaction.rb:239:in `block in within_new_transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/connection_adapters/abstract/transaction.rb:236:in `within_new_transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/connection_adapters/abstract/database_statements.rb:267:in `transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/transactions.rb:212:in `transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1343:in `ddl_transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1291:in `execute_migration_in_transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1263:in `block in migrate_without_lock'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1262:in `each'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1262:in `migrate_without_lock'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1210:in `block in migrate'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1363:in `with_advisory_lock'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1210:in `migrate'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1036:in `up'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1011:in `migrate'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/tasks/database_tasks.rb:172:in `migrate'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/railties/databases.rake:60:in `block (2 levels) in <main>'

Caused by:
NoMethodError: undefined method `allow_signup' for #<Permission:0x0000000009c679e8>
/usr/local/rvm/gems/ruby-2.6.6/gems/activemodel-5.2.4.4/lib/active_model/attribute_methods.rb:430:in `method_missing'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/relation/delegation.rb:71:in `each'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/relation/delegation.rb:71:in `each'
/home/xxx/zammad/app/models/role.rb:227:in `reject'
/home/xxx/zammad/app/models/role.rb:227:in `check_default_at_signup_permissions'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:426:in `block in make_lambda'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:198:in `block (2 levels) in halting'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:606:in `block (2 levels) in default_terminator'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:605:in `catch'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:605:in `block in default_terminator'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:199:in `block in halting'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:513:in `block in invoke_before'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:513:in `each'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:513:in `invoke_before'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:131:in `run_callbacks'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:816:in `_run_update_callbacks'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/callbacks.rb:350:in `_update_record'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/timestamp.rb:114:in `_update_record'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/persistence.rb:705:in `create_or_update'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/callbacks.rb:342:in `block in create_or_update'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:132:in `run_callbacks'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:816:in `_run_save_callbacks'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/callbacks.rb:342:in `create_or_update'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/persistence.rb:308:in `save!'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/validations.rb:52:in `save!'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/transactions.rb:315:in `block in save!'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/transactions.rb:387:in `block in with_transaction_returning_status'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/connection_adapters/abstract/database_statements.rb:265:in `transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/transactions.rb:212:in `transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/transactions.rb:385:in `with_transaction_returning_status'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/transactions.rb:315:in `save!'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/suppressor.rb:48:in `save!'
/home/xxx/zammad/db/migrate/20200527000000_agent_customer.rb:8:in `block in change'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/relation/delegation.rb:71:in `each'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/relation/delegation.rb:71:in `each'
/home/xxx/zammad/db/migrate/20200527000000_agent_customer.rb:6:in `change'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:814:in `exec_migration'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:798:in `block (2 levels) in migrate'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:797:in `block in migrate'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:416:in `with_connection'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:796:in `migrate'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:977:in `migrate'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1292:in `block in execute_migration_in_transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1343:in `block in ddl_transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/connection_adapters/abstract/database_statements.rb:267:in `block in transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/connection_adapters/abstract/transaction.rb:239:in `block in within_new_transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/connection_adapters/abstract/transaction.rb:236:in `within_new_transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/connection_adapters/abstract/database_statements.rb:267:in `transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/transactions.rb:212:in `transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1343:in `ddl_transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1291:in `execute_migration_in_transaction'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1263:in `block in migrate_without_lock'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1262:in `each'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1262:in `migrate_without_lock'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1210:in `block in migrate'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1363:in `with_advisory_lock'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1210:in `migrate'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1036:in `up'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1011:in `migrate'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/tasks/database_tasks.rb:172:in `migrate'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/railties/databases.rake:60:in `block (2 levels) in <main>'
Tasks: TOP => db:migrate
(See full trace by running task with --trace)
```

### Steps to reproduce the behavior:

* Install Zammad 3.3 and upgrade to 3.6

PS: Possible solution would be to merge `db/migrate/20200527000000_agent_customer.rb` and `db/migrate/20201112000000_role_signup_column_fix.rb` into one file.

**Comment by dukex** [comment]
My workaround was run just the 20201112000000

```
$ zammad run rake db:migrate:up VERSION=20201112000000
```

And after run the migration again
```
$ zammad run rake db:migrate
```

**Comment by hydrapolic** [comment]
Thanks, same for upgrade from 3.4.0 -> 3.6.0. Applying commit https://github.com/zammad/zammad/commit/a41de24e27ccb6c9bf15f95e66e5484e5bd8c80f fixed the situation.

**Comment by djmaze** [comment]
Just stumbled upon this as well. Shouldn't there have been a 3.6.1 release at some point, since this bug breaks upgrades?

Instead I am seeing a 3.7.0 release, but there is no mention and no download available for that.

**Comment by rolfschmidt** [comment]
Hi @djmaze ,

can you add some more details about from which version you upgraded to and also the target version? The bug should be fixed? Does your migration file contain the following line?

https://github.com/zammad/zammad/blob/9a3a289ac7493590d23f9522077ca4d7fdc0c748/db/migrate/20200527000000_agent_customer.rb#L8

**Comment by djmaze** [comment]
I updated to the latest version 3.6.0 which is available via ftp. (TBH, this was a *fresh* test installation using docker, but the problem is there either way.)

Not surprisingly, the line is not there since the fix [is not in 3.6.0](https://github.com/zammad/zammad/blob/3.6.0/db/migrate/20200527000000_agent_customer.rb).

**Comment by rolfschmidt** [comment]
> 
> 
> I updated to the latest version 3.6.0 which is available via ftp. (TBH, this was a _fresh_ test installation using docker, but the problem is there either way.)
> 
> Not surprisingly, the line is not there since the fix [is not in 3.6.0](https://github.com/zammad/zammad/blob/3.6.0/db/migrate/20200527000000_agent_customer.rb).

Okay the 3.6.1 is the zammad-latest release. If you have a the possibility try it again with this one:

https://ftp.zammad.com/zammad-latest.tar.gz
https://ftp.zammad.com/zammad-latest.tar.bz2
https://ftp.zammad.com/zammad-latest.zip

or if you are using git (use the stable branch).

3.7.0 isnt released/stable yet.

afaik the 3.6.1 tag will have a synchronized release with the next major release (3.7). We have a little delay because of some new heavy features.

**Comment by MrGeneration** [comment]
Sorry but I'm locking the conversation to contributors only to reduce the noise on this repo.
Above was rather a technical question that should life on our community board: https://community.zammad.org.
